api/authentication.py

```diff
@@ -1,160 +1,414 @@
+    if not isinstance(sub := data.get("sub", None), str) or not sub:
+        return IntrospectionError("NoSubject", status_code=status_code, data=fxa_data)
+
+    return IntrospectionResponse(data=cast(FxaIntrospectData, data))
+
+
+def introspect_token_or_raise(
+    token: str, use_cache: bool = True
+) -> IntrospectionResponse:
+    """
+    Introspect a Mozilla account OAuth token, to get data like the FxA UID.
+
+    If anything goes wrong, raise an exception.
+    """
+    default_cache_timeout = 60
+
+    fxa_resp: IntrospectionResponse | IntrospectionError | None = None
+    if use_cache:
+        fxa_resp = load_introspection_result_from_cache(cache, token)
+    if fxa_resp is None:
+        fxa_resp = introspect_token(token)
+
+    # If the response is an error, raise an exception
+    if isinstance(fxa_resp, IntrospectionError):
+        if not fxa_resp.from_cache:
+            fxa_resp.save_to_cache(cache, token, default_cache_timeout)
+        fxa_resp.raise_exception()
+
+    if not fxa_resp.from_cache:
+        cache_timeout = max(default_cache_timeout, fxa_resp.cache_timeout)
+        fxa_resp.save_to_cache(cache, token, cache_timeout)
+    return fxa_resp
+
+
+class FxaTokenAuthentication(TokenAuthentication):
+    """
+    Implement authentication with a Mozilla Account bearer token.
+
+    This is passed by Firefox for the Accounts user. Unlike DRF's
+    TokenAuthentication, this is not generated by Relay. Instead, it
+    needs to be validated by Mozilla Accounts to get the FxA ID.
+    """
+
+    keyword = "Bearer"
+    relay_user_required = True
+
+    def authenticate(
+        self, request: Request
+    ) -> None | tuple[User | AnonymousUser, IntrospectionResponse]:
+        """
+        Try to authenticate with a Accounts bearer token.
+
+        If successful, it returns a tuple (user, token), which can be accessed at
+        request.user and request.auth. Also, request.successful_authenticator will be
+        an instance of this class.
+
+        If it fails, it raises an APIException with a status code:
+        * 503 Service Unavailable - The introspect API request failed, or had bad data
+        * 401 Authentication Failed - The introspect API says the account is inactive
+        * 403 Forbidden - The introspect API returns an active account, but the
+          matching Relay user is required.
+
+        If the authentication header is not an Accounts bearer token, it returns None
+        to skip to the next authentication method.
+        """
         method = request.method
-        if method in ["POST", "DELETE", "PUT"]:
-            use_cache = False
-            if method == "POST" and request.path == "/api/v1/relayaddresses/":
-                use_cache = True
-        fxa_uid = get_fxa_uid_from_oauth_token(token, use_cache)
+        path = request.path
+        self.use_cache = (method == "POST" and path == "/api/v1/relayaddresses/") or (
+            method not in ("POST", "DELETE", "PUT")
+        )
```

Comment:
suggestion (non-blocking): This logic seems harder to follow than nested `if` statements. Can we revert that? (I also don't remember why we force `use_cache=True` when `POST`ing to `/api/v1/relayaddresses` but I'm sure there's a reason.)